x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SystemIndicesUpgradeIT.java

```diff
@@ -0,0 +1,194 @@
+        for (int i = 0; i < propertyPath.size() - 1; i++) {
+            Object o = properties.get(propertyPath.get(i));
+            assertThat(o, instanceOf(Map.class));
+            properties = (Map<String, Object>) o;
+        }
+        return String.valueOf(properties.get(propertyPath.get(propertyPath.size() - 1)));
+    }
+
+    private void upgradeSystemIndices() throws Exception {
+        String upgradeUser = "upgrade_user";
+        String upgradeUserPassword = "x-pack-test-password";
+        createRole("upgrade_role", "irrelevant");
+        createUser(upgradeUser, upgradeUserPassword, "upgrade_role");
+
+        try (RestClient upgradeUserClient = getClient(upgradeUser, upgradeUserPassword)) {
+            boolean upgradeRequired = Version.fromString(UPGRADE_FROM_VERSION).before(SystemIndices.NO_UPGRADE_REQUIRED_VERSION);
+            String expectedStatus = (upgradeRequired) ? "MIGRATION_NEEDED" : "NO_MIGRATION_NEEDED";
+
+            assertThat(
+                XContentHelper.convertToMap(
+                    JsonXContent.jsonXContent,
+                    upgradeUserClient.performRequest(new Request("GET", "/_migration/system_features")).getEntity().getContent(),
+                    false
+                ).get("migration_status"),
+                equalTo(expectedStatus)
+            );
+
+            if (upgradeRequired) {
+                Request upgradeRequest = new Request("POST", "/_migration/system_features");
+                Response upgradeResponse = upgradeUserClient.performRequest(upgradeRequest);
+                assertOK(upgradeResponse);
+                assertBusy(() -> {
+                    Response featureResponse = upgradeUserClient.performRequest(new Request("GET", "/_migration/system_features"));
+                    assertThat(
+                        XContentHelper.convertToMap(JsonXContent.jsonXContent, featureResponse.getEntity().getContent(), false)
+                            .get("migration_status"),
+                        equalTo("NO_MIGRATION_NEEDED")
+                    );
+                });
+            }
+        }
+    }
+
+    private void createUser(String name, String password, String role) throws IOException {
+        Request request = new Request("PUT", "/_security/user/" + name);
+        request.setJsonEntity("{ \"password\": \"" + password + "\", \"roles\": [ \"" + role + "\"] }");
+        assertOK(adminClient().performRequest(request));
+    }
+
+    private void createRole(String name, String indexOrDataStream) throws IOException {
+        Request request = new Request("PUT", "/_security/role/" + name);
+        request.setJsonEntity(
+            "{ \"cluster\": [\"cluster:admin/migration/post_system_feature\", \"cluster:admin/migration/get_system_feature\"],"
+                + " \"indices\": [ { \"names\" : [ \""
+                + indexOrDataStream
+                + "\"], \"privileges\": [ \"manage\" ] } ] }"
```

Comment:
This part, and the passed-in indexOrDataStream could be removed, since the only permissions needed for this to work are the cluster-level permissions.